design/baremetal-operator/raid-disk-controller-naming.md

```diff
@@ -0,0 +1,173 @@
+
+### Goals
+
+The primary goal is to reach consensus on one naming convention to depict
+disk and RAID controller names in the baremetal-host ``hardware-raid``
+section.
+
+Afterwards, implementation of the same will be done within the
+``baremetal-operator`` by extending the ``baremetal-host`` specification.
+
+### Non-Goals
+
+This specification does not deal with ``software-raid`` and extension for it.
+
+## Proposal
+
+The crux of the proposal is the naming convention that is to be adopted
+for interfacing with the user. Currently, we have the following options
+
+### Controller Naming
+
+For controllers that are installed specifically for the operating system,
+for example the [BOSS Card](1) by DellEMC, they are almost always used with
+two disks/drives in ``RAID-1``, and can be refered to as
+``primary-controller``. Similar method can be adopted to address such
+controllers by other vendors.
+
+Other controllers that are installed in the server can be used in any order,
+and thus can be named ``secondary-controller-x`` where x can be an integer
+greater than or equal to 1, e-g ``secondary-controller-1``,
+``secondary-controller-2`` and so on.
+
+### Physical Disk/Drive Naming
+
+With physical disks, the primary issue is disk enumeration. We can address
+it this way:
+
+- Start from the front bay, and count drives in the order that they are
+  listed by vendors. For instance, vendors would list them as XXXXX-1-XXX,
+  XXXXX-2-XXX and so on. We can use these 1 and 2 to specify ordering and
+  provide a simpler naming convention as ``Disk-1``, ``Disk-2`` and so on.
+- The drives inserted in the backplane can be counted _afterwards_ using
+  the same convention.
+
+This will provide the user with a simpler way of listing controllers
+and disks and it should suffice for a significant majority of use-cases.
+
+### User Stories
+
+The following user stories appertain to the proposal in question
+
+#### Story 1
+
+As an operator, I should be able to focus on defining my RAID configurations
+without having to deal with intricacies of individual vendor's naming
+conventions.
+
+#### Story 2
+
+As an operator, I'd like to be able to specify the physical disks I want
+to use for my RAID configuration. Or, in case of multiple controllers, I
+should be able to choose from them in defining my ``hardware-raid``
+configuration, or both.
+
+## Design Details
+
+- The CRD spec will have to be extended to add fields for ``physical_disks``
+  and ``controller`` under the ``hardware_raid`` section.
+- The provisioner will then be extended to process these fields.
+- There needs to be a map to transform user-interfacing names to the
+  vendor-specific names (like in case of ``firmware`` config).
+- The provisioner logic will perform lookups, and make Ironic API calls
+  accordingly.
+- The status field will mirror these fields after successful configuration.
+  This is to provide the operator with the information regarding current config.
+
```

Comment:
would you mind giving a full example of the CRD here please ?

Reply:
Sure, let me add those

Reply:
@maelk I've added the sample CRD into the latest patch. The naming convention is up for debate and I'd really like to have input on that. Thank you for your time!

Reply:
Is hardwareVolumes the correct term to use in the yaml? Perhaps logicalVolumes is a better term to use.

Reply:
It's "hardwareVolumes" as in "volumes managed by a hardware RAID controller", as opposed to software RAID. Both hardware and software RAID produce logical volumes so I don't think that name describes what we want, but I agree the current naming is awkward. I don't have any good suggestions though.

Reply:
Dah yes you're right. Let's come up with some names. How about ``hardwareRAIDVolume``? I think that does it well.

Reply:
I've updated the naming to ``hardwareRAIDVolume(s)``.